File: Source/Aladdin.CAPI.KZ.Java/src/aladdin/capi/kz/rsa/KeyFactory.java
package aladdin.capi.kz.rsa;
import aladdin.asn1.*; 
import aladdin.asn1.kz.*; 
import aladdin.capi.*; 

///////////////////////////////////////////////////////////////////////////
// Параметры ключа RSA
///////////////////////////////////////////////////////////////////////////
public class KeyFactory extends aladdin.capi.ansi.rsa.KeyFactory
{
    // параметры ключа
    private final aladdin.capi.ansi.rsa.IParameters parameters; 
    
    // конструктор
    public KeyFactory(String keyOID) { super(keyOID); 
    
        // в зависимости от идентификатора ключа
        if (keyOID.equals(OID.GAMMA_KEY_RSA_1024) || 
            keyOID.equals(OID.GAMMA_KEY_RSA_1024_XCH))
        {
            // указать параметры ключа
            parameters = new aladdin.capi.ansi.rsa.Parameters(1024, null); 
        }
        else if (keyOID.equals(OID.GAMMA_KEY_RSA_1536) || 
                 keyOID.equals(OID.GAMMA_KEY_RSA_1536_XCH))
        {
            // указать параметры ключа
            parameters = new aladdin.capi.ansi.rsa.Parameters(1536, null); 
        }
        else if (keyOID.equals(OID.GAMMA_KEY_RSA_2048) || 
                 keyOID.equals(OID.GAMMA_KEY_RSA_2048_XCH))
        {
            // указать параметры ключа
            parameters = new aladdin.capi.ansi.rsa.Parameters(2048, null); 
        }
        else if (keyOID.equals(OID.GAMMA_KEY_RSA_3072) || 
                 keyOID.equals(OID.GAMMA_KEY_RSA_3072_XCH))
        {
            // указать параметры ключа
            parameters = new aladdin.capi.ansi.rsa.Parameters(3072, null); 
        }
        else if (keyOID.equals(OID.GAMMA_KEY_RSA_4096) || 
                 keyOID.equals(OID.GAMMA_KEY_RSA_4096_XCH))
        {
            // указать параметры ключа
            parameters = new aladdin.capi.ansi.rsa.Parameters(4096, null); 
        }
        // при ошибке выбросить исключение
        else throw new UnsupportedOperationException(); 
    } 
    // способ использования ключа
    @Override public KeyUsage getKeyUsage() 
    { 
        // для специальных ключей
        if (keyOID().equals(OID.GAMMA_KEY_RSA_1024) || 
            keyOID().equals(OID.GAMMA_KEY_RSA_1536) || 
            keyOID().equals(OID.GAMMA_KEY_RSA_2048))
        {
            // указать способ использования ключа
            return new KeyUsage(
                KeyUsage.DIGITAL_SIGNATURE | KeyUsage.CERTIFICATE_SIGNATURE | 
                KeyUsage.CRL_SIGNATURE     | KeyUsage.NON_REPUDIATION
            ); 
        }
        // вызвать базовую функцию
        return super.getKeyUsage(); 
    }
    // закодировать параметры
    @Override public IEncodable encodeParameters(
        aladdin.capi.IParameters parameters) { return Null.INSTANCE; }
    
    // параметры алгоритма
    @Override public final aladdin.capi.ansi.rsa.IParameters 
        decodeParameters(IEncodable encoded) { return parameters; }
}

File: Source/Aladdin.CAPI.KZ/CAPI/KZ/RSA/KeyFactory.cs
using System;

namespace Aladdin.CAPI.KZ.RSA
{
    ///////////////////////////////////////////////////////////////////////////
    // Параметры ключа RSA
    ///////////////////////////////////////////////////////////////////////////
    public class KeyFactory : CAPI.ANSI.RSA.KeyFactory
    {
        // параметры ключа 
        private CAPI.ANSI.RSA.IParameters parameters; 

        // конструктор
        public KeyFactory(string keyOID) : base(keyOID) 
        {
            // в зависимости от идентификатора ключа
            if (keyOID == ASN1.KZ.OID.gamma_key_rsa_1024 || 
                keyOID == ASN1.KZ.OID.gamma_key_rsa_1024_xch)
            {
                // указать параметры ключа
                parameters = new ANSI.RSA.Parameters(1024, null); 
            }
            // в зависимости от идентификатора ключа
            else if (keyOID == ASN1.KZ.OID.gamma_key_rsa_1536 || 
                     keyOID == ASN1.KZ.OID.gamma_key_rsa_1536_xch)
            {
                // вернуть параметры алгоритма
                parameters = new ANSI.RSA.Parameters(1536, null); 
            }
            // в зависимости от идентификатора ключа
            else if (keyOID == ASN1.KZ.OID.gamma_key_rsa_2048 || 
                     keyOID == ASN1.KZ.OID.gamma_key_rsa_2048_xch)
            {
                // вернуть параметры алгоритма
                parameters = new ANSI.RSA.Parameters(2048, null); 
            }
            // в зависимости от идентификатора ключа
            else if (keyOID == ASN1.KZ.OID.gamma_key_rsa_3072 || 
                     keyOID == ASN1.KZ.OID.gamma_key_rsa_3072_xch)
            {
                // вернуть параметры алгоритма
                parameters = new ANSI.RSA.Parameters(3072, null); 
            }
            // в зависимости от идентификатора ключа
            else if (keyOID == ASN1.KZ.OID.gamma_key_rsa_4096 || 
                     keyOID == ASN1.KZ.OID.gamma_key_rsa_4096_xch)
            {
                // вернуть параметры алгоритма
                parameters = new ANSI.RSA.Parameters(4096, null); 
            }
            // при ошибке выбросить исключение
            else throw new NotSupportedException(); 
        }
	    // способ использования ключа
	    public override KeyUsage GetKeyUsage() 
        { 
            // для специальных ключей
            if (KeyOID == ASN1.KZ.OID.gamma_key_rsa_1024 || 
                KeyOID == ASN1.KZ.OID.gamma_key_rsa_1536 || 
                KeyOID == ASN1.KZ.OID.gamma_key_rsa_2048)
            {
            // указать способ использования ключа
            return KeyUsage.DigitalSignature | KeyUsage.CertificateSignature | 
                   KeyUsage.CrlSignature     | KeyUsage.NonRepudiation;       
            }
            // вызвать базовую функцию
            return base.GetKeyUsage(); 
        }
        // закодировать параметры
        public override ASN1.IEncodable EncodeParameters(
            CAPI.IParameters parameters) { return ASN1.Null.Instance; }

        // раскодировать параметры
        public override CAPI.IParameters 
            DecodeParameters(ASN1.IEncodable encoded) { return parameters; }
    }
}
